File: array/118_Pascal'sTriangle.py
# coding: utf8


"""
    题目链接: https://leetcode.com/problems/pascals-triangle/description.
    题目描述:

    Given numRows, generate the first numRows of Pascal's triangle.

    For example,
        given numRows = 5,

        Return

        [
             [1],
            [1,1],
           [1,2,1],
          [1,3,3,1],
         [1,4,6,4,1]
        ]

"""


class Solution(object):
    def generate(self, numRows):
        """
        :type numRows: int
        :rtype: List[List[int]]
        """
        if numRows == 0:
            return []
        elif numRows == 1:
            return [[1]]
        elif numRows == 2:
            return [[1], [1, 1]]

        triangles = [[1], [1, 1]]
        for row in range(2, numRows):
            triangle = [1]
            for column in range(1, row):
                triangle.append(triangles[row - 1][column - 1] + triangles[row - 1][column])
            triangle.append(1)
            triangles.append(triangle)

        return triangles

File: Array/llb/118. Pascal's Triangle.cpp
/*
问题描述：
Given numRows, generate the first numRows of Pascal's triangle.

For example, given numRows = 5,
Return

[
     [1],
    [1,1],
   [1,2,1],
  [1,3,3,1],
 [1,4,6,4,1]
]
*/

/*思路：
每行的两边都是1，中间则有上一行的j和j-1相加而成
*/

class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        //思路：每行两边的都是1，中间都是上一行的j和j-1之和
        // vector<vector<int>> resultVec(numRows);
        // for (int i=0; i<numRows; i++) {
        //     resultVec[i].resize(i+1);
        //     resultVec[i][0] = 1;
        //     resultVec[i][i] = 1;
        //     for (int j=1; j<i; j++) {
        //         resultVec[i][j] = resultVec[i-1][j-1] + resultVec[i-1][j];
        //     }
        // }
        // return resultVec;
        
        vector<vector<int>> resultVec(numRows);
        for (int i=0; i<numRows; i++) {
            vector<int> tempVec(i+1);
            tempVec[0] = tempVec[i] = 1;            
            for (int j=1; j<i; j++) {
                tempVec[j] = resultVec[i-1][j-1] + resultVec[i-1][j];
            }
            resultVec[i] = tempVec;
        }
        return resultVec;
    }
};

